src/spaces/continuous.jl

```diff
@@ -0,0 +1,442 @@
+accelerate nearest neighbor functions like [`nearby_ids`](@ref).
+All dimensions in `extent` must be completely divisible by `spacing` (i.e. no
+fractional remainder).
+Your agent positions (field `pos`) must be of type `NTuple{D, <:Real}`.
+
+The keyword argument `update_vel!` is a **function**, `update_vel!(agent, model)` that updates
+the agent's velocity **before** the agent has been moved, see [`move_agent!`](@ref).
+You can of course change the agents' velocities
+during the agent interaction, the `update_vel!` functionality targets spatial force
+fields acting on the agents individually (e.g. some magnetic field).
+By default no update is done this way.
+If you use `update_vel!`, the agent type must have a field `vel::NTuple{D, <:Real}`.
+
+The keyword `periodic = true` configures whether the space is periodic or not.
+
+**Note:** if your model requires linear algebra operations for which tuples are not supported,
+a performant solution is to convert between Tuple and SVector using
+[StaticArrays.jl](https://github.com/JuliaArrays/StaticArrays.jl)
+as follows: `s = SVector(t)` and back with `t = Tuple(s)`.
+"""
+function ContinuousSpace(
+        extent::NTuple{D,X},
+        spacing = min(extent)/10.0;
+        update_vel! = defvel,
+        periodic = true,
+    ) where {D, X<:Real}
+    @assert all(extent./spacing .== floor.(extent./spacing)) "All dimensions in `extent` must be completely divisible by `spacing`"
+    s = GridSpace(floor.(Int, extent ./ spacing), periodic = periodic, metric = :euclidean)
+    Z = X <: AbstractFloat ? X : Float64
+    return ContinuousSpace(s, update_vel!, size(s), Z(spacing), Z.(extent))
+end
+
+function random_position(model::ABM{A,<:ContinuousSpace}) where {A}
+    map(dim->rand()*dim, model.space.extent)
+end
+
+pos2cell(pos::Tuple, model::ABM) = floor.(Int, pos ./ model.space.spacing) .+ 1
+pos2cell(a::AbstractAgent, model::ABM) = pos2cell(a.pos, model)
+function cell_center(pos::NTuple{D, F}, model) where {D, F}
+    ε = model.space.spacing
+    (pos2cell(pos, model) .- 1) .* ε .+ ε/2
+end
+distance_from_cell_center(pos, model::ABM) =
+distance_from_cell_center(pos, cell_center(pos, model))
+function distance_from_cell_center(pos::Tuple, center::Tuple)
+    sqrt(sum(abs2.(pos .- center)))
+end
+
+
+function add_agent_to_space!(
+        a::A,
+        model::ABM{A,<:ContinuousSpace},
+    ) where {A<:AbstractAgent}
+    push!(model.space.grid.s[pos2cell(a, model)...], a.id)
+    return a
+end
+
+function remove_agent_from_space!(
+        a::A,
+        model::ABM{A,<:ContinuousSpace},
+    ) where {A<:AbstractAgent}
+    prev = model.space.grid.s[pos2cell(a, model)...]
+    ai = findfirst(i -> i == a.id, prev)
+    deleteat!(prev, ai)
+    return a
+end
+
+function move_agent!(
+        a::A,
+        pos::Tuple,
+        model::ABM{A,<:ContinuousSpace{D,periodic}},
+    ) where {A<:AbstractAgent,D,periodic}
+    remove_agent_from_space!(a, model)
+    if periodic
```

Comment:
Should we dispatch explicit values of periodic here rather than using the `if`?

Reply:
No, it doesn't make a difference. The `if` gets pruned in compilation.